File: android/app/src/main/java/com/ayuvo/health/actions/CoachActionTools.kt
package com.ayuvo.health.actions

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import java.util.UUID

/** A write Coach suggested; nothing changes until the user taps Confirm on its card. */
data class ActionProposal(
    val actionId: String,
    val title: String,
    val params: Map<String, Any?>,
    val lines: List<String>,
    val ai: Boolean,
    val id: String = UUID.randomUUID().toString()
)

/**
 * Coach's view of the action catalog (docs/actions.md, AI Coach): `coach_mode: read` actions are
 * tools named by their `coach_tool`, run through the same [ActionExecutor] as every other surface
 * with [ActionSource.COACH]; `coach_mode: propose` actions are reachable only through
 * [PROPOSE_TOOL], which validates and records an [ActionProposal] without writing anything.
 * Medication and goal changes are never proposable (catalog rule, checked by the contract).
 */
class CoachActionTools(private val executor: ActionExecutor) {
    private val catalog = executor.catalog
    private val readByTool = catalog.coachReadActions.associateBy { it.coachTool!! }

    /** Proposals collected during one Coach turn, in call order. */
    val proposals: MutableList<ActionProposal> = mutableListOf()

    val names: List<String> = readByTool.keys.toList() + PROPOSE_TOOL

    fun handles(name: String): Boolean = name in readByTool || name == PROPOSE_TOOL

    fun description(name: String): String? = when (name) {
        PROPOSE_TOOL -> "Suggest a change to the user's Ayuvo log (for example logging water, a weight, a food with known calories, a workout set, or starting or ending a fast). " +
            "It does NOT change anything: the user sees a card and decides. Only call it when the user asked for the change. " +
            "action_id is one of: " + catalog.coachProposeActions.joinToString("; ") { "${it.id} (${it.summary})" } +
            ". params_json is a JSON object of that action's parameters: " +
            catalog.coachProposeActions.joinToString("; ") { a -> "${a.id}: " + a.params.joinToString(", ") { "${it.name} (${it.summary})" } }
        else -> readByTool[name]?.let { "${it.summary} (Ayuvo action ${it.id})" }
    }

    /** JSON schema text for [name]'s parameters. */
    fun schemaJson(name: String): String? {
        if (name == PROPOSE_TOOL) {
            val ids = catalog.coachProposeActions.map { it.id }
            return ActionValues.toJson(linkedMapOf(
                "type" to "object",
                "properties" to linkedMapOf(
                    "action_id" to linkedMapOf("type" to "string", "enum" to ids, "description" to "The Ayuvo action to propose."),
                    "params_json" to linkedMapOf("type" to "string", "description" to "The action's parameters as a JSON object, e.g. {\"amount\": 500, \"unit\": \"ml\"}.")
                ),
                "required" to listOf("action_id")
            )).toString()
        }
        val spec = readByTool[name] ?: return null
        val props = LinkedHashMap<String, Any?>()
        for (p in spec.params) {
            val prop = LinkedHashMap<String, Any?>()
            when (p.type) {
                ParamType.NUMBER -> prop["type"] = "number"
                ParamType.INTEGER -> prop["type"] = "integer"
                ParamType.ENUM -> { prop["type"] = "string"; prop["enum"] = catalog.enums[p.enumName].orEmpty() }
                else -> prop["type"] = "string"
            }
            if (p.min != null && p.unitParam == null) prop["minimum"] = p.min
            if (p.max != null && p.unitParam == null) prop["maximum"] = p.max
            prop["description"] = p.summary
            props[p.name] = prop
        }
        val schema = linkedMapOf<String, Any?>("type" to "object", "properties" to props)
        spec.params.filter { it.required }.map { it.name }.takeIf { it.isNotEmpty() }?.let { schema["required"] = it }
        return ActionValues.toJson(schema).toString()
    }

    suspend fun execute(name: String, args: Map<String, Any?>): String {
        if (name == PROPOSE_TOOL) return propose(args)
        val spec = readByTool[name] ?: return error("unknown_tool")
        return try {
            executor.perform(ActionRequest(spec.id, args.filterValues { it != null }, ActionSource.COACH)).toJson().toString()
        } catch (e: ActionException) {
            error(e.code.raw, e.param)
        }
    }

    private suspend fun propose(args: Map<String, Any?>): String {
        val id = args["action_id"] as? String ?: return error("missing_param", "action_id")
        val params: Map<String, Any?> = when (val raw = args["params_json"] ?: args["params"]) {
            null -> emptyMap()
            is Map<*, *> -> raw.entries.associate { it.key.toString() to it.value }
            is String -> if (raw.isBlank()) emptyMap() else runCatching {
                @Suppress("UNCHECKED_CAST")
                ActionValues.fromJson(Json.parseToJsonElement(raw) as JsonObject) as Map<String, Any?>
            }.getOrElse { return error("bad_type", "params_json") }
            else -> return error("bad_type", "params_json")
        }
        val spec = catalog.action(id)
        if (spec == null || spec.coachMode != CoachMode.PROPOSE) return error("not_allowed")
        return when (val v = executor.validate(ActionRequest(id, params, ActionSource.COACH))) {
            is ValidationResult.Failed -> error(v.code.raw, v.param)
            is ValidationResult.Ok -> {
                val proposal = ActionProposal(id, spec.title, v.action.params, ActionResultText.paramLines(spec, v.action.params), v.action.ai)
                proposals += proposal
                ActionValues.toJson(linkedMapOf(
                    "ok" to true,
                    "status" to "proposed",
                    "note" to "Nothing was changed. The user now sees a card with Confirm and Dismiss; tell them to review it."
                )).toString()
            }
        }
    }

    private fun error(code: String, param: String? = null): String =
        ActionValues.toJson(linkedMapOf("ok" to false, "error" to code, "param" to param)).toString()

    companion object {
        const val PROPOSE_TOOL = "propose_action"
    }
}

File: ios/calorietracker/Actions/CoachActionTools.swift
import SwiftUI

/// Coach side of the action catalog (docs/actions.md §Coach): `coach_mode: read` actions become extra
/// read-only Coach tools, and `propose_action` lets Coach suggest a `coach_mode: propose` write that
/// only runs when the user taps Confirm on the card under the reply. Coach never writes by itself and
/// can never propose medication or goal changes (the catalog marks those `coach_mode: none`).
nonisolated struct CoachActionProposal: Identifiable, Equatable, Sendable {
    let id = UUID()
    let validation: ActionValidation
    let title: String
    let summary: String
}

/// Collects proposals made during one Coach turn.
@MainActor
final class CoachActionProposalSink {
    private(set) var proposals: [CoachActionProposal] = []

    func add(_ proposal: CoachActionProposal) {
        // A model that repeats the same proposal in one turn gets one card.
        guard !proposals.contains(where: { $0.validation == proposal.validation }) else { return }
        proposals.append(proposal)
    }
}

extension CoachTools {
    static let proposeToolName = "propose_action"

    /// Catalog read tools (`coach_tool` names), in catalog order.
    static var actionReadToolNames: [String] {
        ActionCatalog.shared.actions.filter { $0.coachMode == "read" }.compactMap(\.coachTool)
    }

    static var proposableActionIDs: [String] {
        ActionCatalog.shared.actions.filter { $0.coachMode == "propose" }.map(\.id)
    }

    /// Local diary tools follow the food source (no extra consent); proposals also need a UI to confirm.
    var actionToolNames: [String] {
        guard availableToolNames.contains("get_data_summary") else { return [] }
        return Self.actionReadToolNames + (actionProposals == nil ? [] : [Self.proposeToolName])
    }

    /// Everything advertised to the provider this turn.
    /// Catalog tools go after the diary / health tools and before the records and medications
    /// contract tools, so the existing tool order is unchanged.
    var allToolNames: [String] {
        let base = availableToolNames
        let contract = Set(Self.recordsToolNames + Self.medicationToolNames)
        let split = base.firstIndex { contract.contains($0) } ?? base.count
        return Array(base[..<split]) + actionToolNames + Array(base[split...])
    }

    static func isActionTool(_ name: String) -> Bool {
        name == proposeToolName || actionReadToolNames.contains(name)
    }

    static func actionFor(tool name: String) -> ActionCatalog.Action? {
        ActionCatalog.shared.actions.first { $0.coachTool == name }
    }

    static func description(for name: String) -> String {
        if name == proposeToolName {
            let actions = ActionCatalog.shared.actions.filter { $0.coachMode == "propose" }.map { action in
                let params = action.params.map { param -> String in
                    var text = param.name
                    if let name = param.enumName, let values = ActionCatalog.shared.enums[name] { text += " (\(values.joined(separator: "|")))" }
                    return param.required ? text + "*" : text
                }.joined(separator: ", ")
                return "\(action.id): \(action.summary) Params: \(params.isEmpty ? "none" : params)."
            }.joined(separator: " ")
            return "Suggest a change to the user's Ayuvo log. Nothing is saved: the user sees a card with Confirm and decides. "
                + "Only propose when the user asked to log or start/stop something. * = required. Actions: \(actions)"
        }
        if let action = actionFor(tool: name) {
            return "\(action.summary) (Ayuvo action \(action.id); read-only, local data.)"
        }
        return toolDescriptions[name] ?? ""
    }

    static func schema(for name: String) -> [String: Any] {
        if name == proposeToolName {
            return [
                "type": "object",
                "properties": [
                    "action_id": ["type": "string", "enum": proposableActionIDs, "description": "Which Ayuvo action to propose"],
                    "params_json": ["type": "string", "description": "The action's parameters as a JSON object string, e.g. {\"amount\": 500, \"unit\": \"ml\"}"],
                ],
                "required": ["action_id", "params_json"],
            ]
        }
        if let action = actionFor(tool: name) {
            var properties: [String: Any] = [:]
            for param in action.params {
                var property: [String: Any] = ["description": param.summary]
                switch param.type {
                case .number: property["type"] = "number"
                case .integer: property["type"] = "integer"
                case .enum:
                    property["type"] = "string"
                    if let name = param.enumName, let values = ActionCatalog.shared.enums[name] { property["enum"] = values }
                case .string, .metric, .entity: property["type"] = "string"
                }
                properties[param.name] = property
            }
            var schema: [String: Any] = ["type": "object", "properties": properties]
            let required = action.params.filter(\.required).map(\.name)
            if !required.isEmpty { schema["required"] = required }
            return schema
        }
        return parameterSchema(for: name)
    }

    func executeActionTool(name: String, arguments: [String: Any]) async -> String {
        let executor = ActionExecutor.shared
        if name == Self.proposeToolName {
            guard let actionID = arguments["action_id"] as? String,
                  executor.catalog.action(actionID)?.coachMode == "propose" else {
                return ActionJSON.error(.invalid(code: "not_allowed", param: "action_id"))
            }
            var params: [String: Any] = [:]
            let rawParams = arguments["params_json"] ?? arguments["params"]
            if let object = rawParams as? [String: Any] {
                params = object
            } else if let text = rawParams as? String, let data = text.data(using: .utf8),
                      let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] {
                params = object
            }
            switch executor.validate(actionID, ActionRawValue.params(params), source: .coach) {
            case .failure(let error):
                return ActionJSON.error(error, actionID: actionID)
            case .success(let validation):
                let summary = executor.summary(of: validation)
                actionProposals?.add(CoachActionProposal(validation: validation, title: executor.catalog.action(actionID)?.title ?? actionID, summary: summary))
                return ActionJSON.text([
                    "proposed": true, "action": actionID, "summary": summary,
                    "note": "Shown to the user as a card with a Confirm button. Nothing is saved until they tap Confirm; tell them to confirm below.",
                ])
            }
        }
        guard let action = Self.actionFor(tool: name) else { return ActionJSON.error(.invalid(code: "unknown_action", param: nil)) }
        do {
            return try await executor.run(action.id, ActionRawValue.params(arguments), source: .coach).jsonText
        } catch let error as ActionError {
            return ActionJSON.error(error, actionID: action.id)
        } catch {
            return ActionJSON.error(.unavailable(error.localizedDescription), actionID: action.id)
        }
    }
}

/// Confirm card under a Coach reply for one proposed action.
struct CoachActionProposalCard: View {
    let proposal: CoachActionProposal
    let outcome: String?
    let isRunning: Bool
    let onConfirm: () -> Void
    let onDismiss: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Label(proposal.title, systemImage: "checkmark.seal")
                .font(.system(.footnote, design: .rounded, weight: .semibold))
                .foregroundStyle(.secondary)
            Text(proposal.summary)
                .font(.system(.subheadline, design: .rounded))
            if let outcome {
                Text(outcome)
                    .font(.system(.footnote, design: .rounded))
                    .foregroundStyle(.secondary)
                    .accessibilityIdentifier("coach.proposal.outcome")
            } else {
                HStack(spacing: 10) {
                    Button {
                        onConfirm()
                    } label: {
                        Text("Confirm").frame(maxWidth: .infinity)
                    }
                    .buttonStyle(.borderedProminent)
                    .disabled(isRunning)
                    .accessibilityIdentifier("coach.proposal.confirm")
                    Button(role: .cancel) {
                        onDismiss()
                    } label: {
                        Text("Not now").frame(maxWidth: .infinity)
                    }
                    .buttonStyle(.bordered)
                    .disabled(isRunning)
                }
                .font(.system(.footnote, design: .rounded, weight: .semibold))
            }
        }
        .padding(12)
        .background(.ultraThinMaterial, in: RoundedRectangle(cornerRadius: 14, style: .continuous))
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("coach.proposal")
    }
}
